### PR TITLE
fix(backend): Fix incorrect cross-origin detection behind AWS CloudFront

### PR DESCRIPTION
**Pull request by flybayer**
## Type of change

- [x] 🐛 Bug fix

## Packages affected

- [x] `@clerk/backend`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

AWS CloudFront always sets the origin header protocol to `http`, even when the request is `https`. It's impossible to change. This causes Clerk behind CloudFront to incorrectly detect the request as cross-origin.

This fixes that, and adds a test case for it.

### CODE REVIEW

**Comment by dimkl** [comment]
hello @flybayer thank you for the contribution :)
Could you describe a bit your deployed application architecture?
Eg are you using `NextJS Static Exports` to deploy a NextJS application as a SPA using CloudFront + S3?

**Comment by flybayer** [comment]
@dimkl Sure, this is a Next.js application running on ECS via `next start`. It's fronted by CloudFront with an Application Load Balancer in between CloudFront and ECS.

The CloudFront is passing host and origin headers, the issue remains.

**Comment by dimkl** [comment]
@flybayer Could you set `debug: true` in `authMiddleware()` and provide the generated logs?

**Comment by flybayer** [comment]
@dimkl unfortunately I don't have a project I can test that on. This has been reported by a few of our users on https://www.flightcontrol.dev

**Comment by ebg1223** [comment]
i can help test. Having this issue! @dimkl

**Comment by ebg1223** [comment]
@flybayer 
I was able to get a flightcontrol deployment running using your changes, but I believe your condition may be backwards.
Had to change to this to get working:
// AWS CloudFront always sets FINAL protocol to http, and it's impossible to change
if (finalURL.host === originURL.host && finalURL.protocol === 'http:' && originURL.protocol === 'https:') 

It seems final protocol is http and origin is https. Let me know if this is a valid change.

File: packages/backend/src/util/request.ts [hunk 1]
```diff
@@ -17,8 +17,19 @@ export function checkCrossOrigin({
...
+  const finalURLstring = buildOrigin({ forwardedProto, forwardedHost, protocol: originURL.protocol, host });
+  if (!finalURLstring) {
+    return false;
+  }
+  const finalURL = new URL(finalURLstring);
+  if (finalURL.origin === originURL.origin) {
+    return false;
+  }
+  // AWS CloudFront always sets origin protocol to http, and it's impossible to change
+  if (finalURL.host === originURL.host && finalURL.protocol === 'https:' && originURL.protocol === 'http:') {
```
Comment (flybayer) [inline packages/backend/src/util/request.ts:29]:
```suggestion
  if (finalURL.host === originURL.host && finalURL.protocol === 'http:' && originURL.protocol === 'https:') {
```

**Comment by flybayer** [comment]
Thanks @ebg1223, I've updated the code based on your testing

**Comment by flybayer** [comment]
@ebg1223 are you able to share those logs?

**Comment by ebg1223** [comment]
> @ebg1223 are you able to share those logs?
Yes- I just need to prep an environment. I patched this package successfully and now flightcontrol is serving production traffic. Should be able to share tomorrow.

**Comment by ebg1223** [comment]
This is a log of a failed post request. Get requests are processed successfully.

Identifying details omitted, but if you need anything, let me know, and I can dm.

```
2023-09-06T22:22:58.028-07:00 | INFO: Clerk: The request to / is being redirected because there is no signed-in user, and the path is not included in `ignoredRoutes` or `publicRoutes`. To prevent this behavior, choose one of:
-- | --
  | 2023-09-06T22:22:58.028-07:00 | 1. To make the route accessible to both signed in and signed out users, add "/" to the `publicRoutes` array passed to authMiddleware
  | 2023-09-06T22:22:58.028-07:00 | 2. To prevent Clerk authentication from running at all, pass `ignoredRoutes: ["/((?!api\|trpc))(_next\|.+\..+)(.*)", "/"]` to authMiddleware
  | 2023-09-06T22:22:58.028-07:00 | 3. Pass a custom `afterAuth` to authMiddleware, and replace Clerk's default behavior of redirecting unless a route is included in publicRoutes
  | 2023-09-06T22:22:58.028-07:00 | For additional information about middleware, please visit https://clerk.com/docs/nextjs/middleware
  | 2023-09-06T22:22:58.028-07:00 | (This log only appears in development mode, or if `debug: true` is passed to authMiddleware)
  | 2023-09-06T22:22:58.028-07:00 | Clerk debug start :: authMiddleware
  | 2023-09-06T22:22:58.028-07:00 | -- URL debug, {
  | 2023-09-06T22:22:58.028-07:00 | "url": "http://localhost:8150/",
  | 2023-09-06T22:22:58.028-07:00 | "method": "GET",
  | 2023-09-06T22:22:58.028-07:00 | "headers": "{omitted sentry details}"x-forwarded-for\":\"::ffff:10.192.11.79\",\"x-forwarded-host\":\"{OMITTED_SERVER_IP}:8150\",\"x-forwarded-port\":\"8150\",\"x-forwarded-proto\":\"http\",\"x-invoke-path\":\"\",\"x-invoke-query\":\"\",\"x-middleware-invoke\":\"1\"}",
  | 2023-09-06T22:22:58.028-07:00 | "nextUrl": "http://localhost:8150/",
  | 2023-09-06T22:22:58.028-07:00 | "clerkUrl": "http://{OMITTED_SERVER_IP}:8150/"
  | 2023-09-06T22:22:58.028-07:00 | }
  | 2023-09-06T22:22:58.028-07:00 | -- Options debug, {
  | 2023-09-06T22:22:58.028-07:00 | "signInUrl": "/sign-in",
  | 2023-09-06T22:22:58.028-07:00 | "debug": true,
  | 2023-09-06T22:22:58.028-07:00 | "beforeAuth": false,
  | 2023-09-06T22:22:58.028-07:00 | "afterAuth": false
  | 2023-09-06T22:22:58.028-07:00 | }
  | 2023-09-06T22:22:58.028-07:00 | -- {
  | 2023-09-06T22:22:58.028-07:00 | "auth": "{\"sessionClaims\":null,\"sessionId\":null,\"session\":null,\"userId\":null,\"user\":null,\"actor\":null,\"orgId\":null,\"orgRole\":null,\"orgSlug\":null,\"organization\":null,\"isPublicRoute\":false,\"isApiRoute\":false}",
  | 2023-09-06T22:22:58.028-07:00 | "debug": {
  | 2023-09-06T22:22:58.028-07:00 | "signInUrl": "/sign-in",
  | 2023-09-06T22:22:58.028-07:00 | "debug": true,
  | 2023-09-06T22:22:58.028-07:00 | "request": {
  | 2023-09-06T22:22:58.028-07:00 | "sourcePage": "/middleware",
  | 2023-09-06T22:22:58.028-07:00 | "experimental_clerkUrl": "http://{OMITTED_SERVER_IP}:8150/"
  | 2023-09-06T22:22:58.028-07:00 | },
  | 2023-09-06T22:22:58.028-07:00 | "apiKey": "",
  | 2023-09-06T22:22:58.028-07:00 | "secretKey": "sk_live",
  | 2023-09-06T22:22:58.028-07:00 | "audience": "",
  | 2023-09-06T22:22:58.028-07:00 | "apiUrl": "https://api.clerk.dev",
  | 2023-09-06T22:22:58.028-07:00 | "apiVersion": "v1",
  | 2023-09-06T22:22:58.028-07:00 | "frontendApi": "{OMITTED}",
  | 2023-09-06T22:22:58.028-07:00 | "proxyUrl": "",
  | 2023-09-06T22:22:58.028-07:00 | "publishableKey": "{OMITTED}",
  | 2023-09-06T22:22:58.028-07:00 | "isSatellite": false,
  | 2023-09-06T22:22:58.028-07:00 | "domain": "",
  | 2023-09-06T22:22:58.028-07:00 | "jwtKey": "",
  | 2023-09-06T22:22:58.028-07:00 | "searchParams": {},
  | 2023-09-06T22:22:58.028-07:00 | "host": "{OMITTED_SERVER_IP}:8150",
  | 2023-09-06T22:22:58.028-07:00 | "forwardedHost": "{OMITTED_SERVER_IP}:8150",
  | 2023-09-06T22:22:58.028-07:00 | "forwardedPort": "8150",
  | 2023-09-06T22:22:58.028-07:00 | "forwardedProto": "http",
  | 2023-09-06T22:22:58.028-07:00 | "userAgent": "ELB-HealthChecker/2.0",
  | 2023-09-06T22:22:58.028-07:00 | "status": "signed-out",
  | 2023-09-06T22:22:58.028-07:00 | "reason": "cookie-and-uat-missing",
  | 2023-09-06T22:22:58.028-07:00 | "message": ""
  | 2023-09-06T22:22:58.028-07:00 | }
  | 2023-09-06T22:22:58.028-07:00 | }
  | 2023-09-06T22:22:58.028-07:00 | -- {
  | 2023-09-06T22:22:58.028-07:00 | "mergedHeaders": {OMITTED}
  | 2023-09-06T22:22:58.028-07:00 | }
  | 2023-09-06T22:22:58.028-07:00 | -- Final response is redirect, following redirect
 
```

**Comment by ebg1223** [comment]
@flybayer I see the latest commit added this check:
forwardedProto === 'https', but according to these logs, forwardedProto": "http". Any reason you added that?

The patch I am running is as follow, and is working, but want to make sure I am not missing anything!
`
+  const finalURLstring = buildOrigin({ forwardedProto, forwardedHost, protocol: originURL.protocol, host });
+  if (!finalURLstring) {
+    return false;
+  }
+  const finalURL = new URL(finalURLstring);
+  if (finalURL.origin === originURL.origin) {
+    return false;
+  }
+  // AWS CloudFront always sets FINAL protocol to http, and it's impossible to change
+  if (finalURL.host === originURL.host && finalURL.protocol === 'http:' && originURL.protocol === 'https:') {
+    return false;
+  }
+  return true;
 }
+
`

**Comment by flybayer** [comment]
Ok I removed that, but have to change invert the logic for 2 existing tests 🤔

**Comment by flybayer** [comment]
@jescalan could you remove the `blocked` label? Afaik I addressed the requirements now

**Review by dimkl — COMMENTED** [review]
Hello @flybayer 
Based on the changes in this PR, the current and expected behaviour of cross-origin validation breaks and it makes every request with an `http` protocol to be valid for origins with `https` which is not valid based on the definition found in MDN.

> Two URLs have the same origin if the [protocol](https://developer.mozilla.org/en-US/docs/Glossary/Protocol), [port](https://developer.mozilla.org/en-US/docs/Glossary/Port) (if specified), and [host](https://developer.mozilla.org/en-US/docs/Glossary/Host) are the same for both. You may see this referenced as the "scheme/host/port tuple", or just "tuple".

MDN: https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy#definition_of_an_origin

**Comment by flybayer** [comment]
@dimkl do you have any idea on a way to fix this for AWS without changing that existing behavior?

@ebg1223, you have any idea either?

**Comment by dimkl** [comment]
Hello @flybayer 😄 
i've managed to deploy an application using flightcontrol.dev and checked the generated CloudFront configuration.
I've tried some things and i think that we could fix the issue with some changes in `CloudFront / Distributions /
<DistributionID> / Edit origin` but failed to make it work:
Screenshot 2023-09-20 at 19 48 57

Despite that, I think we could use one of the following approaches to resolve this:
-  enable `CloudFront-Forwarded-Proto` in CloudFront and use it `@clerk/backend` package [ref](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/adding-cloudfront-headers.html#cloudfront-headers-other)
- allow bypassing the `x-forwarded-proto` from application's code by passing `forwardedProto`  to authenticate request (eg for nextjs it should be passed to `authMiddleware` options and a `@ts-ignore` will be required)    THIS IS A HACK - not suggested
- find some other way to config CloudFront -> ELB communication to pass the correct value to the x-forwarded-proto

I would go for the 3rd option and even ask the AWS support and if there is no solution i would go with the 1st solution.

**Comment by flybayer** [comment]
Thanks @dimkl! 

Unfortunately option 3 is not possible, but option 1 will work!

I've updated this PR to take that approach

**Comment by ebg1223** [comment]
@flybayer thank you for doing this!! I agree option 1 is the only practical solution.

**Comment by dimkl** [comment]
@flybayer thank you for the appying those changes. I think we are good to go.
Some final notes:
- could add a test to assert `CloudFront-Forwarded-Proto` case
- rebase and resolve the current conflicts

I will approve and merge this change when the above notes are resolved :)

**Comment by dimkl** [comment]
@flybayer is it okay if i cherry-pick the commits to a new PR and proceed with merging it because there are conflicts and I will be difficult to ping you to rebase the branch until we can finally merge it?
Also, I will take a look if I can add any test or not.
Thank you for the patience and for the contribution.

**Comment by flybayer** [comment]
@dimkl yes go ahead!

**Comment by dimkl** [comment]
Closing in favor of : https://github.com/clerkinc/javascript/pull/1817

**Comment by clerk-cookie** [comment]
This PR has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.

**Pull request closed without merging**